### PR TITLE
Update all multifile datasets to use ncrcat

### PR DESCRIPTION
**Pull request by xylar**
Some obs. and preprocessed datasets were being opened with `xarray.open_mfdataset()` function (via the the "generalized reader" in MPAS-Analysis), which sometimes led to "too many open files" errors.

### CODE REVIEW

**Comment by xylar** [comment]
@milenaveneziani, the plots before and after this fix should be identical.  There should be output in the `timeSeries/preprocessed` output folder that contains the data combined with ncrcat.

These fixes are meant to remove all use of `generalized_reader.open_mutlifile_dataset()` because I'm getting "too many open files" errors in #327.

**Comment by milenaveneziani** [comment]
Hi @xylar: I am testing this on edison, using this conda environment:
`e3sm_unified_1.2.0_py2.7_nox`
which I believe is the latest e3sm-unified.

I am getting the following error:
```
Plotting SST time series...
  Load SST data...
  Load in SST for a preprocesses reference run...
running: ncrcat -4 --record_append --no_tmp_fl / ... c /global/project/projectdirs/acme/milena/20180129.DECKv1b_piControl.ne30_oEC.edison_transects_new/timeseries/preprocessed/sst.nc
ncrcat: ERROR nco_sng_sntz() reports character '*' from unsanitized user-input string "*" is not on whitelist of acceptable characters. For security purposes NCO restricts the set of characters appearing in user input, including filenames, to: "abcdefghijklmnopqrstuvwxyzABCDEFGHIJKLMNOPQRSTUVWXYZ1234567890_-.@ :%/". NB: This restriction was first imposed in NCO 4.7.3 (February, 2018), and may cause breakage of older workflows. Please contact NCO if you have a real-world use-case that shows why the character '*' should be white-listed. HINT: Re-try command after replacing transgressing characters with innocuous characters.

analysis task timeSeriesSST failed during run
Traceback (most recent call last):
  File "/global/u2/m/milena/MPAS-git-repositories/MPAS-Analysis/mpas_analysis/shared/analysis_task.py", line 317, in run
    self.run_task()
  File "/global/u2/m/milena/MPAS-git-repositories/MPAS-Analysis/mpas_analysis/ocean/time_series_sst.py", line 209, in run_task
    outFileName, logger=self.logger)
  File "/global/u2/m/milena/MPAS-git-repositories/MPAS-Analysis/mpas_analysis/shared/time_series/time_series.py", line 108, in combine_time_series_with_ncrcat
    ' '.join(args))
CalledProcessError: Command 'ncrcat -4 --record_append --no_tmp_fl / g l o b a l / p r o j e c t / p r o j e c t d i r s / a c m e / A C M E v 0 _ l o w r e s / B 1 8 5 0 C 5 _ n e 3 0 _ v 0 . 4 / o c n / p o s t p r o c e s s i n g / S S T . B 1 8 5 0 C 5 _ n e 3 0 _ v 0 . 4 . y e a r * . n c /global/project/projectdirs/acme/milena/20180129.DECKv1b_piControl.ne30_oEC.edison_transects_new/timeseries/preprocessed/sst.nc' returned non-zero exit status 1

Execution time: 0:00:00.16
```

Does it have to do with the nco version?

**Comment by xylar** [comment]
Looks like a bug on my part and not an error in netcdf. I'll look into it tomorrow. Thanks very much for testing.

**Comment by xylar** [comment]
@milenaveneziani, this turns out to be a python 2 vs. 3 issue (again!).    Sorry about that!

**Comment by xylar** [comment]
Tests on a QU240wLI run passed with both python 2 and python 3

**Comment by milenaveneziani** [comment]
hmm, another pesky error. This is on cori, and I've been getting it even after pruging:
```
Plotting sea-ice area and volume time series...
  Load sea-ice data...
  Make NH plots...
running: ncrcat -4 --record_append --no_tmp_fl /global/project/projectdirs/acme/observations/SeaIce/IceArea_timeseries/iceAreaNH_climo.nc ... /global/project/projectdirs/acme/observations/SeaIce/IceArea_timeseries/iceAreaNH_climo.nc /global/project/projectdirs/acme/milena/20180129.DECKv1b_piControl.ne30_oEC.edison_transects_new/timeseries/obs/iceAreaNH.nc
running: ncrcat -4 --record_append --no_tmp_fl /global/project/projectdirs/acme/observations/SeaIce/PIOMAS/PIOMASvolume_monthly_climo.nc ... /global/project/projectdirs/acme/observations/SeaIce/PIOMAS/PIOMASvolume_monthly_climo.nc /global/project/projectdirs/acme/milena/20180129.DECKv1b_piControl.ne30_oEC.edison_transects_new/timeseries/obs/iceVolumeNH.nc
running: ncrcat -4 --record_append --no_tmp_fl /global/project/projectdirs/acme/ACMEv0_lowres/B1850C5_ne30_v0.4/ice/postprocessing/icearea.B1850C5_ne30_v0.4.year0001.nc ... /global/project/projectdirs/acme/ACMEv0_lowres/B1850C5_ne30_v0.4/ice/postprocessing/icearea.B1850C5_ne30_v0.4.year0169.nc /global/project/projectdirs/acme/milena/20180129.DECKv1b_piControl.ne30_oEC.edison_transects_new/timeseries/preprocessed/iceAreaNH.nc
running: ncrcat -4 --record_append --no_tmp_fl /global/project/projectdirs/acme/ACMEv0_lowres/B1850C5_ne30_v0.4/ice/postprocessing/icevol.B1850C5_ne30_v0.4.year0001.nc ... /global/project/projectdirs/acme/ACMEv0_lowres/B1850C5_ne30_v0.4/ice/postprocessing/icevol.B1850C5_ne30_v0.4.year0169.nc /global/project/projectdirs/acme/milena/20180129.DECKv1b_piControl.ne30_oEC.edison_transects_new/timeseries/preprocessed/iceVolumeNH.nc
analysis task timeSeriesSeaIceAreaVol failed during run
Traceback (most recent call last):
  File "/global/u2/m/milena/MPAS-git-repositories/MPAS-Analysis/mpas_analysis/shared/analysis_task.py", line 317, in run
    self.run_task()
  File "/global/u2/m/milena/MPAS-git-repositories/MPAS-Analysis/mpas_analysis/sea_ice/time_series.py", line 449, in run_task
    yearStrideXTicks=yearStrideXTicks)
  File "/global/u2/m/milena/MPAS-git-repositories/MPAS-Analysis/mpas_analysis/shared/plot/plotting.py", line 149, in timeseries_analysis_plot
    minDays.append(mean.Time.min())
  File "/global/project/projectdirs/acme/software/anaconda_envs/cori/base/envs/e3sm_unified_1.2.0_py2.7_nox/lib/python2.7/site-packages/xarray/core/common.py", line 22, in wrapped_func
    skipna=skipna, allow_lazy=True, **kwargs)
  File "/global/project/projectdirs/acme/software/anaconda_envs/cori/base/envs/e3sm_unified_1.2.0_py2.7_nox/lib/python2.7/site-packages/xarray/core/dataarray.py", line 1361, in reduce
    var = self.variable.reduce(func, dim, axis, keep_attrs, **kwargs)
  File "/global/project/projectdirs/acme/software/anaconda_envs/cori/base/envs/e3sm_unified_1.2.0_py2.7_nox/lib/python2.7/site-packages/xarray/core/variable.py", line 1267, in reduce
    axis=axis, **kwargs)
  File "/global/project/projectdirs/acme/software/anaconda_envs/cori/base/envs/e3sm_unified_1.2.0_py2.7_nox/lib/python2.7/site-packages/xarray/core/duck_array_ops.py", line 320, in f
    return func(values, axis=axis, **kwargs)
  File "/global/project/projectdirs/acme/software/anaconda_envs/cori/base/envs/e3sm_unified_1.2.0_py2.7_nox/lib/python2.7/site-packages/xarray/core/dataarray.py", line 1361, in reduce
    var = self.variable.reduce(func, dim, axis, keep_attrs, **kwargs)
  File "/global/project/projectdirs/acme/software/anaconda_envs/cori/base/envs/e3sm_unified_1.2.0_py2.7_nox/lib/python2.7/site-packages/xarray/core/variable.py", line 1267, in reduce
    axis=axis, **kwargs)
  File "/global/project/projectdirs/acme/software/anaconda_envs/cori/base/envs/e3sm_unified_1.2.0_py2.7_nox/lib/python2.7/site-packages/xarray/core/duck_array_ops.py", line 320, in f
    return func(values, axis=axis, **kwargs)
  File "/global/project/projectdirs/acme/software/anaconda_envs/cori/base/envs/e3sm_unified_1.2.0_py2.7_nox/lib/python2.7/site-packages/xarray/core/duck_array_ops.py", line 49, in f
    return getattr(module, name)(*args, **kwargs)
ValueError: numpy.nanmin raises on a.size==0 and axis=None; So Bottleneck too.

Execution time: 0:00:11.54
```

**Comment by xylar** [comment]
Thanks @milenaveneziani.  I'll make sure I can reproduce this and try to figure out what's going on as soon as I can.  It looks like something is empty for some reason.

**Comment by milenaveneziani** [comment]
ah, I only see iceAreaNH and iceVolumeNH in the obs directory.

**Comment by milenaveneziani** [comment]
so I am missing `iceAreaSH.nc` for some reason.

**Comment by xylar** [comment]
That's very odd.  How did we not see this before?

**Comment by milenaveneziani** [comment]
no clue. Let me try again.

**Comment by milenaveneziani** [comment]
still no `iceaAreaSH.nc`.. I am puzzled.

**Comment by xylar** [comment]
@milenaveneziani, here are the files it's actually looking for:
https://github.com/MPAS-Dev/MPAS-Analysis/blob/develop/mpas_analysis/config.default#L1568

Could it be that you changed them in the custom config file but you've got them in the wrong section because of changes I made in #408?

**Comment by xylar** [comment]
Ignore what I just wrote.  It's about obs and you're looking at preprocessed data, which wasn't affected by #408.  Sorry.

**Comment by milenaveneziani** [comment]
no, my error is about the obs, I think. It's not making the file `obs/iceAreaSH.nc`.
I am trying to read my full log file more carefully, give me a min.

**Comment by xylar** [comment]
I'm trying it out myself right now on cori.

**Comment by milenaveneziani** [comment]
I now think it is the `preprocessed` data, but I don't understand why. The data is there if I `ls` it..

**Comment by xylar** [comment]
I was just able to run successfully on cori with this config file:
```
[runs]
## options related to the run to be analyzed and reference runs to be
## compared against

# mainRunName is a name that identifies the simulation being analyzed.
mainRunName = 20180129.DECKv1b_piControl.ne30_oEC.edison

# preprocessedReferenceRunName is the name of a reference run that has been
# preprocessed to compare against (or None to turn off comparison).  Reference
# runs of this type would have preprocessed results because they were not
# performed with MPAS components (so they cannot be easily ingested by
# MPAS-Analysis)
preprocessedReferenceRunName = B1850C5_ne30_v0.4

# config file for a reference run to which this run will be compared.  The
# analysis should have already been run to completion once with this config
# file, so that the relevant MPAS climatologies already exist and have been
# remapped to the comparison grid.  Leave this option commented out if no
# reference run is desired.
# referenceRunConfigFile = /path/to/config/file

[execute]
## options related to executing parallel tasks

# the number of parallel tasks (1 means tasks run in serial, the default)
parallelTaskCount = 4

# the parallelism mode in ncclimo ("serial" or "bck")
# Set this to "bck" (background parallelism) if running on a machine that can
# handle 12 simultaneous processes, one for each monthly climatology.
ncclimoParallelMode = bck

[input]
## options related to reading in the results to be analyzed

# directory containing model results
baseDirectory = /global/cscratch1/sd/golaz/ACME_simulations/20180129.DECKv1b_piControl.ne30_oEC.edison

# subdirectory containing restart files
runSubdirectory = run
# subdirectory for ocean history files
oceanHistorySubdirectory = archive/ocn/hist
# subdirectory for sea ice history files
seaIceHistorySubdirectory = archive/ice/hist

# names of ocean and sea ice meshes (e.g. oEC60to30v3, oQU240v3, oRRS30to10v3, etc.)
mpasMeshName = oEC60to30v3

# Directory for mapping files (if they have been generated already). If mapping
# files needed by the analysis are not found here, they will be generated and
# placed in the output mappingSubdirectory
mappingDirectory = /global/project/projectdirs/acme/mpas_analysis/mapping

# names of namelist and streams files, either a path relative to baseDirectory
# or an absolute path.
oceanNamelistFileName = run/mpas-o_in
oceanStreamsFileName = run/streams.ocean
seaIceNamelistFileName = run/mpas-cice_in
seaIceStreamsFileName = run/streams.cice

[output]
## options related to writing out plots, intermediate cached data sets, logs,
## etc.

# directory where analysis should be written
baseDirectory = /global/cscratch1/sd/xylar/analysis/ncrcat_obs/20180129.DECKv1b_piControl.ne30_oEC.edison

# provide an absolute path to put HTML in an alternative location (e.g. a web
# portal)
# htmlSubdirectory = /global/project/projectdirs/acme/www/USERNAME/RUNNAME
htmlSubdirectory = html

# a list of analyses to generate.  Valid names can be seen by running:
#   ./run_mpas_analysis --list
# This command also lists tags for each analysis.
# Shortcuts exist to generate (or not generate) several types of analysis.
# These include:
#   'all' -- all analyses will be run
#   'all_<tag>' -- all analysis with a particular tag will be run
#   'all_<component>' -- all analyses from a given component (either 'ocean'
#                        or 'seaIce') will be run
#   'only_<component>', 'only_<tag>' -- all analysis from this component or
#                                       with this tag will be run, and all
#                                       analysis for other components or
#                                       without the tag will be skipped
#   'no_<task_name>' -- skip the given task
#   'no_<component>', 'no_<tag>' -- in analogy to 'all_*', skip all analysis
#                                   tasks from the given compoonent or with
#                                   the given tag.  Do
#                                      ./run_mpas_analysis --list
#                                   to list all task names and their tags
# an equivalent syntax can be used on the command line to override this
# option:
#    ./run_mpas_analysis config.analysis --generate \
#         all,no_ocean,all_timeSeries
# All tasks with tag "landIceCavities" are disabled because this run did not
# include land-ice cavities.
generate = ['all', 'no_landIceCavities']

[climatology]
## options related to producing climatologies, typically to compare against
## observations and previous runs

# the first year over which to average climatalogies
startYear = 5
# the last year over which to average climatalogies
endYear = 10

[timeSeries]
## options related to producing time series plots, often to compare against
## observations and previous runs

# start and end years for timeseries analysis. Using out-of-bounds values
#   like start_year = 1 and end_year = 9999 will be clipped to the valid range
#   of years, and is a good way of insuring that all values are used.
startYear = 1
endYear = 10

[index]
## options related to producing nino index.

# start and end years for the nino 3.4 analysis.  Using out-of-bounds values
#   like start_year = 1 and end_year = 9999 will be clipped to the valid range
#   of years, and is a good way of insuring that all values are used.
# For valid statistics, index times should include at least 30 years
startYear = 1
endYear = 10

[oceanObservations]
## options related to ocean observations with which the results will be compared

# directory where ocean observations are stored
baseDirectory = /global/project/projectdirs/acme/observations/Ocean/
sstSubdirectory = SST
sssSubdirectory = SSS
mldSubdirectory = MLD
ninoSubdirectory = Nino
mhtSubdirectory = MHT


[oceanPreprocessedReference]
## options related to preprocessed ocean reference run with which the results
## will be compared (e.g. a POP, CESM or ACME v0 run)

# directory where ocean reference simulation results are stored
baseDirectory = /global/project/projectdirs/acme/ACMEv0_lowres/B1850C5_ne30_v0.4/ocn/postprocessing

[seaIceObservations]
## options related to sea ice observations with which the results will be
## compared

# directory where sea ice observations are stored
baseDirectory = /global/project/projectdirs/acme/observations/SeaIce

[seaIcePreprocessedReference]
## options related to preprocessed sea ice reference run with which the results
## will be compared (e.g. a CICE, CESM or ACME v0 run)

# directory where ocean reference simulation results are stored
baseDirectory =  /global/project/projectdirs/acme/ACMEv0_lowres/B1850C5_ne30_v0.4/ice/postprocessing

[timeSeriesSeaIceAreaVol]
## options related to plotting time series of sea ice area and volume

# plot on polar plot
polarPlot = False

[regions]
## options related to ocean regions used in several analysis modules

# Directory for region mask files
regionMaskDirectory = /global/project/projectdirs/acme/mpas_analysis/region_masks
```
and this set of commands:
```
. /global/project/projectdirs/acme/software/anaconda_envs/${NERSC_HOST}/base/etc/profile.d/conda.sh
conda activate e3sm_unified_1.2.0_py2.7_nox
./run_mpas_analysis --purge --generate=timeSeriesSeaIceAreaVol config.20180129.DECKv1b_piControl.ne30_oEC.edison
```

**Comment by xylar** [comment]
I rebased onto develop to get rid of the display error I was getting because of #412.  You may want to do a hard reset.

**Comment by xylar** [comment]
If you notice any differences, let me know.  I'm happy to re-test with exactly your setup.

File: mpas_analysis/sea_ice/time_series.py [hunk 1]
```diff
@@ -260,25 +260,6 @@ def run_task(self):  # {{{
...
-                                                   calendar=calendar).year
-            if yearStart <= preprocessedYearEnd:
-                dsPreprocessedTimeSlice = \
-                    dsPreprocessed.sel(Time=slice(timeStart, timeEnd))
-            else:
-                self.logger.warning('Preprocessed time series ends before the '
-                                    'timeSeries startYear and will not be '
-                                    'plotted.')
-                preprocessedReferenceRunName = 'None'
-
```
Comment (milenaveneziani) [inline mpas_analysis/sea_ice/time_series.py]: I am looking at the changes more closely, to try to understand what is going on in my weird ice timeseries error.
So, I noticed this piece above: is it being replaced by something else?

Comment (xylar) [reply mpas_analysis/sea_ice/time_series.py]: I think it's totally redundant to the code below.  Maybe I'm missing something.

File: mpas_analysis/sea_ice/time_series.py [hunk 2]
```diff
@@ -389,11 +385,14 @@ def run_task(self):  # {{{
                 inFilesPreprocessed = '{}/icevol.{}.year*.nc'.format(
                     preprocessedReferenceDirectory,
                     preprocessedReferenceRunName)
-                dsPreprocessed = open_multifile_dataset(
-                    fileNames=inFilesPreprocessed,
-                    calendar=calendar,
-                    config=config,
-                    timeVariableName='xtime')
```
Comment (xylar) [inline mpas_analysis/sea_ice/time_series.py]: This bit of code did *exactly* the same thing as the block above, I think the block above was a mistake (and didn't accomplish anything).

**Comment by milenaveneziani** [comment]
I am using this config file:
`~milena/MPAS-git-repositories/MPAS-Analysis/config.20180129.DECKv1b_piControl.ne30_oEC.edison`
It would be great if you could run with it on cori, @xylar.

One other thing I noticed, while digging for this, is that we have a `preprocessed/iceVolumeNH.nc` as a product, but that file contains both `icevolume_nh` and `icevolume_sh` variables. Similar argument goes for `preprocessed/iceAreaNH.nc`.

File: mpas_analysis/sea_ice/time_series.py (hunk 1 above)
Comment (xylar) [reply mpas_analysis/sea_ice/time_series.py]: I see.  Maybe what this *does* accomplish is not plotting the reference data if it's out of range.  And now it tries to plot it even if it's out of range?  Maybe your case tests that (because your years are way later than mine)?

Comment (milenaveneziani) [reply mpas_analysis/sea_ice/time_series.py]: I guess I am looking at the piece starting with `if  yearStart`. I remember we put that in there for when the preprocessed timeseries ended before the requested endYear.

Comment (xylar) [reply mpas_analysis/sea_ice/time_series.py]: Let me add this piece back in (with the appropriate changes) and we'll see if it fixes your problem.

Comment (milenaveneziani) [reply mpas_analysis/sea_ice/time_series.py]: yes, exactly, I am running with years=495, I think. way past when the preprocessed data are available. Good catch (both of us :)

**Comment by milenaveneziani** [comment]
Thanks @xylar: I am retesting now and things seem to be working. I'll report back as soon as the test is complete.

**Comment by xylar** [comment]
Very nice detective work, @milenaveneziani.  I think I have fixed both of the issues you pointed out.  There should now only be 1 preprocessed file for area and one for volume (since each contains both hemispheres).  There should be nothing plotted if you run beyond the end of the data set but the volume data gets preprocessed anyway (because that's how we know what the time of the end of the data set is).

**Comment by milenaveneziani** [comment]
yes, that all sounds great.

**Comment by milenaveneziani** [comment]
Results all look good. Thanks @xylar.

**Comment by xylar** [comment]
Thank you!  I really appreciate you doing thorough testing that catches these kinds of strange cases before they become hard to track down. 

I believe this is my last PR before v1.0 (except for a PR that updates versions numbers before we make the release).
